### PR TITLE
massive simplifications

### PR DESCRIPTION
**Pull request by dbmcclain**

### CODE REVIEW

**Comment by dbmcclain** [comment]
I don't know what pure refactoring is? But at any rate, C has the ability to mutate data structures that Rust things are immutable. Hence, I was back and forth on how to ensure that what Rust passes is really okay to have mutated. C will mutate one of the items I pass, just want to be sure it is correct.

I finally settled on the idea that whenever Rust sees a return expression like [0u8; SOME_LENGTH], that it really is freshly constructing a byte vector and returning it, instead of returning some static constant struct. 

And rather than having to constantly type something like that, I invented an internal function within each type Zr, G1, G2, GT, called wv() = working vector, i.e., the actual guts of the tagged data type.

Then finally, I moved things to their own separate files, and added explanatory comments about what each file provides, along with MIT license boilerplate. IOW, I think it is finished, except for a last bit of providing actual unit test code. Pieces were already tested in the example/pbc.rs

**Comment by dbmcclain** [comment]
I rearranged the crypto folder to produce one lib that will contain pbc, curve1174, hash, utils, in preparation for next phase of crypto development. I migrated my existing curve1174.rs over to /examples/curve1174.rs. Pieces of that massive file will be split out and migrated over to the mod.rs file in library folder curve1174. It will share hash and utils with pbc.

File: crypto/src/curve1174/mod.rs [hunk 1]
```diff
@@ -0,0 +1,2306 @@
...
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
```
Comment (rtsisyk) [inline crypto/src/curve1174/mod.rs]: Please fix warnings:

```
warning: unused imports: `ArrayLength`, `GenericArray`
  --> crypto/src/curve1174/mod.rs:30:21
   |
30 | use generic_array::{GenericArray, ArrayLength};
   |                     ^^^^^^^^^^^^  ^^^^^^^^^^^

warning: unused import: `generic_array::typenum::consts::U8`
  --> crypto/src/curve1174/mod.rs:31:5
   |
31 | use generic_array::typenum::consts::U8;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused imports: `Arc`, `Mutex`
  --> crypto/src/curve1174/mod.rs:36:17
   |
36 | use std::sync::{Mutex, Arc};
   |                 ^^^^^  ^^^

warning: unused import: `std::rc::Rc`
  --> crypto/src/curve1174/mod.rs:37:5
   |
37 | use std::rc::Rc;
   |     ^^^^^^^^^^^

warning: unused import: `std::thread`
  --> crypto/src/curve1174/mod.rs:38:5
   |
38 | use std::thread;
   |     ^^^^^^^^^^^

warning: unused import: `std::marker`
  --> crypto/src/curve1174/mod.rs:39:5
   |
39 | use std::marker;
   |     ^^^^^^^^^^^

warning: unused import: `std::vec::*`
  --> crypto/src/curve1174/mod.rs:40:5
   |
40 | use std::vec::*;
   |     ^^^^^^^^^^^

warning: unused import: `utils::*`
  --> crypto/src/curve1174/mod.rs:43:5
   |
43 | use utils::*;
   |     ^^^^^^^^

warning: unused variable: `x`
    --> crypto/src/curve1174/mod.rs:1398:9
     |
1398 | fn gchi(x: FQ51) -> FQ51 {
     |         ^ help: consider using `_x` instead
     |
     = note: #[warn(unused_variables)] on by default

warning: unused variable: `wv`
    --> crypto/src/curve1174/mod.rs:1822:16
     |
1822 |     let WinVec(wv) = w;
     |                ^^ help: consider using `_wv` instead

warning: function is never used: `mul2`
   --> crypto/src/curve1174/mod.rs:316:1
    |
316 | fn mul2(a: &mut [u64; 4]) {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: #[warn(dead_code)] on by default

warning: function is never used: `gadd`
    --> crypto/src/curve1174/mod.rs:1135:1
     |
1135 | fn gadd(x: &FQ51, y: &FQ51, w: &mut FQ51) {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: function is never used: `gsub`
    --> crypto/src/curve1174/mod.rs:1143:1
     |
1143 | fn gsub(x: &FQ51, y: &FQ51, w: &mut FQ51) {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: function is never used: `gdec`
    --> crypto/src/curve1174/mod.rs:1151:1
     |
1151 | fn gdec(x: &FQ51, w: &mut FQ51) {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: function is never used: `gneg`
    --> crypto/src/curve1174/mod.rs:1159:1
     |
1159 | fn gneg(w: &FQ51, x: &mut FQ51) {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: function is never used: `gmul2`
    --> crypto/src/curve1174/mod.rs:1168:1
     |
1168 | fn gmul2(w: &mut FQ51) {
     | ^^^^^^^^^^^^^^^^^^^^^^

warning: function is never used: `gsb2`
    --> crypto/src/curve1174/mod.rs:1177:1
     |
1177 | fn gsb2(x: &FQ51, w: &mut FQ51) {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: function is never used: `gdec2`
    --> crypto/src/curve1174/mod.rs:1394:1
     |
1394 | fn gdec2(x: &mut FQ51) {
     | ^^^^^^^^^^^^^^^^^^^^^^

warning: function is never used: `ecp_neg`
    --> crypto/src/curve1174/mod.rs:1678:1
     |
1678 | fn ecp_neg(qpt: &ECp, ppt: &mut ECp) {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: type `FR_Unscaled` should have a camel case name such as `FrUnscaled`
   --> crypto/src/curve1174/mod.rs:528:1
    |
528 | struct FR_Unscaled(U256);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: #[warn(non_camel_case_types)] on by default

warning: type `FQ_Unscaled` should have a camel case name such as `FqUnscaled`
   --> crypto/src/curve1174/mod.rs:764:1
    |
764 | struct FQ_Unscaled(U256);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: crypto/src/pbc/mod.rs [hunk 2]
```diff
@@ -1,4 +1,5 @@
-// Test PBC Interface with Lynn's PBC Library and our LispPBCIntf glue layer.
+// PBC Crypto for Rust, atop Ben Lynn's PBCliib
```
Comment (rtsisyk) [inline crypto/src/pbc/mod.rs]: Please fix warnings:

```
warning: unused imports: `ArrayLength`, `GenericArray`
  --> crypto/src/pbc/mod.rs:30:21
   |
30 | use generic_array::{GenericArray, ArrayLength};
   |                     ^^^^^^^^^^^^  ^^^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default

warning: unused import: `generic_array::typenum::consts::U8`
  --> crypto/src/pbc/mod.rs:31:5
   |
31 | use generic_array::typenum::consts::U8;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `std::mem`
  --> crypto/src/pbc/mod.rs:34:5
   |
34 | use std::mem;
   |     ^^^^^^^^

warning: unused imports: `Arc`, `Mutex`
  --> crypto/src/pbc/mod.rs:36:17
   |
36 | use std::sync::{Mutex, Arc};
   |                 ^^^^^  ^^^

warning: unused import: `std::rc::Rc`
  --> crypto/src/pbc/mod.rs:37:5
   |
37 | use std::rc::Rc;
   |     ^^^^^^^^^^^

warning: unused import: `std::thread`
  --> crypto/src/pbc/mod.rs:38:5
   |
38 | use std::thread;
   |     ^^^^^^^^^^^

warning: unused import: `std::marker`
  --> crypto/src/pbc/mod.rs:39:5
   |
39 | use std::marker;
   |     ^^^^^^^^^^^

```

File: crypto/src/curve1174/mod.rs [hunk 3]
```diff
@@ -0,0 +1,2306 @@
...
+// -----------------------------------------------------------------
+// type LEV32 represents a 256-bit bignum as a little-endian 32-byte vector
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+struct LEV32([u8;32]);
+
+impl fmt::Display for LEV32 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "LEV32({})", self.nbr_str())
```
Comment (rtsisyk) [inline crypto/src/curve1174/mod.rs]: It is possible to merge nbr_str into `impl fmt::Display for LEV32` and use the standard `to_string()` method everywhere. 

```
Trait std::string::ToString1.0.0[−][src]
[+] Show declaration
[−]

A trait for converting a value to a String.

This trait is automatically implemented for any type which implements the Display trait. As such, ToString shouldn't be implemented directly: Display should be implemented instead, and you get the ToString implementation for free.
```

File: crypto/src/curve1174/mod.rs [hunk 4]
```diff
@@ -0,0 +1,2306 @@
...
+impl Div<FQ51> for FQ51 {
+    type Output = FQ51;
+    fn div(self, x : FQ51) -> FQ51 {
+        let mut tmp = x;
+        ginv(&mut tmp);
+        self * tmp
+    }
+}
+// ---------------------------------------------------------
+// Group primitive operators
```
Comment (rtsisyk) [inline crypto/src/curve1174/mod.rs]: Probably all these function should be moved to FQ51 implementation.

**Pull request closed without merging**